robotoff/prediction/ocr/origin.py

```diff
@@ -0,0 +1,264 @@
+])
+
+COUNTRY_ADJECTIVE_EN_TO_COUNTRY_NAME_EN = {
+    COUNTRIES_IN_ALL_LANGS[country_id]["nationalities"]["en"]:country_id 
+    for country_id in COUNTRIES_IN_ALL_LANGS.keys() 
+    if country_id in COUNTRIES_IN_ALL_LANGS
+    and "nationalities" in COUNTRIES_IN_ALL_LANGS[country_id]
+    and "en" in COUNTRIES_IN_ALL_LANGS[country_id]["nationalities"]
+}
+
+OUTSIDE_WORDS_EN = ["outside"]
+
+
+#Same regex in english. I"m not bilingual in english, so if you find missing terms, missing ways of telling the
+# origin of an ingredient, feel free to add it. You can also add other langages.
+EN_ORIGIN_REGEX = rf"\b(?<!\w)(?P<country_adj>{COUNTRIES_ADJECTIVES_EN_JOINED})? ?(?P<ingredients>(?:(?:{INGREDIENTS_EN_JOINED}) ?(?:and |, ?)? ?)*)? ?(?:100%)? ?(?:is|are|has|have)? ?(?P<negation>'nt|not)? ?(?:{VERBS_EN_JOINED})? ?(?P<in_or_outside>inside|in|from|outside)? ?(?:of)? (?:the)? ?(?P<country>{COUNTRIES_EN_JOINED}) ?(?P<large_origin>(?:several other|several|different?|other|an other) (?:countries|states?|regions?|continents?|origins?))?\b(?!\w)"
+EN_ORIGIN_OCR_REGEX = OCRRegex(
+    re.compile(EN_ORIGIN_REGEX, flags=re.IGNORECASE),
+    field=OCRField.full_text_contiguous,
+    lowercase=True,
+)
+
+
+# Processsing ------------------
+
+EVERY_REGEX_BY_LANG: Dict[str, OCRRegex] = {
+    "fr": FR_ORIGIN_OCR_REGEX,
+    "en": EN_ORIGIN_OCR_REGEX
+}
+GENERAL_WORDS_BY_LANG: Dict[str, List[str]] = {
+    "fr": GENERAL_WORDS_FR,
+    "en": GENERAL_WORDS_EN
+}
+COUNTRY_ADJECTIVE_TO_COUNTRY_NAME_BY_LANG = {
+    "fr": COUNTRY_ADJECTIVE_FR_TO_COUNTRY_NAME_FR,
+    "en": COUNTRY_ADJECTIVE_EN_TO_COUNTRY_NAME_EN
+} 
+
+OUTSIDE_WORDS_BY_LANG = {
+    "fr": OUTSIDE_WORDS_FR,
+    "en": OUTSIDE_WORDS_EN
+}
+
+LARGE_ORIGIN = "large origin" # large origin means several countries, or an unknown country outside of the local area. 
+# (ex: "quinoa comes from outside the E.U", "quinoa has several origins")
+UNKNOW_ORIGIN = "unknow origin"
+ALL_INGREDIENTS = "all ingredients"
+
+def find_origin(content: Union[OCRResult, str]) -> List[Prediction]:
+    ingredients_origins: List[Dict[str, Any]] = []
+    for lang, regex in EVERY_REGEX_BY_LANG.items():
+        text = get_text(content, regex)
+
+        if not text:
+            return []
+
+        for match in regex.regex.finditer(text):
+            origin = extract_origin_from_match(match, lang)
+            if origin == UNKNOW_ORIGIN:
+                continue
+            origin_index = -1
+            for index, ing_ori in enumerate(ingredients_origins):
+                if ing_ori["origin"] == origin:
+                    origin_index = index
+            
+            if origin_index == -1:
+                ingredients_origins.append({
+                    "origin": origin,
+                    "same_for_all_ingredients": True, # True unless group "ingredients" matched
+                    "concerned_ingredients": None
+                })
+                origin_index = len(ingredients_origins) - 1
```

Comment:
why not use a dict (that at the end you will transform to a list of prediction) so that you can index by origin, instead to have to scan the list ? It would make the code more readable.

You could even use a collections.defaultdict:
```python
ingredients_origins: Dict[str, Dict[str, Any]] = collections.defaultdict(
    lambda: {"same_for_all_ingredients": True, "concerned_ingredients": None}
)
```

and here:
```python
origin_data = ingredients_origins[origin]
```

At the end you would use an iterator on items()

Reply:
Yes it's clearly what should be done, I don't why I was scanning the list each time

Reply:
It's always easier to see this kind of thing as a reviewer !